File: daily_problems/2024/11/1120/personal_submission/cf137e_yawn_sean.py
# Submission link: https://codeforces.com/contest/137/submission/292402621
def main():
    s = I()
    vowels = set('aeiouAEIOU')

    n = len(s)

    ans, cnt = 0, 0
    d = Counter()
    d[0] = -1
    d[1] = -1

    cur = 0
    for i in range(n):
        v = -1 if s[i] in vowels else 2
        cur += v
        if cur not in d:
            if v < 0:
                d[cur] = i
            else:
                d[cur - 1] = d[cur - 2]
                d[cur] = d[cur - 2]
        
        v = i
        for j in range(3):
            if cur - j in d:
                v = fmin(v, d[cur - j])
        
        v = i - v
        if v == ans:
            cnt += 1
        elif v > ans:
            ans = v
            cnt = 1

    if ans: print(ans, cnt)
    else: print('No solution')

File: daily_problems/2024/11/1120/personal_submission/cf137e_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/convolution"
// #include "atcoder/dsu"
// #include "atcoder/fenwicktree"
// #include "atcoder/lazysegtree"
// #include "atcoder/math"
// #include "atcoder/maxflow"
// #include "atcoder/mincostflow"
// #include "atcoder/modint"
// #include "atcoder/scc"
// #include "atcoder/segtree"
// #include "atcoder/string"
// #include "atcoder/twosat"
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

typedef array<int, 2> vec;

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    string s;
    cin >> s;

    unordered_set<char> st = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'};

    int n = s.size(), ans = 0, cnt = 0;
    unordered_map<int, int> mp;
    mp[0] = -1;
    mp[1] = -1;

    int cur = 0;

    for (int i = 0; i < n; i ++) {
        int v = (st.find(s[i]) != st.end() ? -1 : 2);
        cur += v;
        if (mp.find(cur) == mp.end()) {
            if (v < 0) mp[cur] = i;
            else mp[cur - 1] = mp[cur - 2], mp[cur] = mp[cur - 2];
        }

        v = i;
        for (int j = 0; j <= 2; j ++) {
            if (mp.find(cur - j) != mp.end()) {
                v = min(v, mp[cur - j]);
            }
        }

        v = i - v;
        if (v == ans) cnt ++;
        else if (v > ans) {
            ans = v;
            cnt = 1;
        }
    }

    if (ans) cout << ans << ' ' << cnt;
    else cout << "No solution";

    return 0;
}